cmd/clusterctl/pkg/client/cluster/components.go

```diff
@@ -86,6 +92,86 @@ func (p *providerComponents) Create(components repository.Components) error {
 	return nil
 }
 
+func (p *providerComponents) Delete(provider clusterctlv1.Provider, forceDeleteNamespace, forceDeleteCRD bool) error {
+	// Fetch all the components belonging to a provider.
+	labels := map[string]string{
+		clusterctlv1.ClusterctlProviderLabelName: provider.Name,
+	}
+	resources, err := p.proxy.ListResources(provider.Namespace, labels)
+	if err != nil {
+		return err
+	}
+
+	// If the CRDs (and the related objects controlled by the provider) should NOT be deleted,
+	// remove CRDs from the list of objects to delete.
+	if !forceDeleteCRD {
+		var resourcesToDelete []unstructured.Unstructured
+		for _, r := range resources {
+			if r.GroupVersionKind().Kind != "CustomResourceDefinition" {
+				resourcesToDelete = append(resourcesToDelete, r)
+			}
+		}
+		resources = resourcesToDelete
+	}
+
+	// If the Namespace where the provider is hosted (and the related objects) should NOT be deleted,
+	// remove the Namespace from the list of objects to delete
+	if !forceDeleteNamespace {
+		var resourcesToDelete []unstructured.Unstructured
+		for _, r := range resources {
+			if r.GroupVersionKind().Kind != "Namespace" {
+				resourcesToDelete = append(resourcesToDelete, r)
+			}
+		}
+		resources = resourcesToDelete
+	} else {
+		// If we are deleting the Namespace where the provider is hosted,
+		// remove namespaced objects from the list of objects to delete (because
+		// everything that is contained in the namespace will be deleted by the Namespace controller)
+		var namespaces []string
+		for _, r := range resources {
+			if r.GroupVersionKind().Kind == "Namespace" {
+				namespaces = append(namespaces, r.GetName())
```

Comment:
Would you expect there to be multiple namespaces coming back from the `p.proxy.ListResources` call? If so, when?

Reply:
This should not happen, because when installing a provider clusterctl checks there is a namespace, but however I thought it was better to make this code more generic. 

If you find this confusing, I'm happy to clean up or to add a note

Reply:
PTAL to how I have refactored the loop. now we are going through resources only 😉